File: WeeklyCode/week6/Merge k Sorted Lists/chenhao.java
public class Solution {
    
    private ListNode findMin(List<ListNode> listOfHeads){
        ListNode minNode=null;
        int minIndex=-1;
        for(int i=0;i<listOfHeads.size();i++){
            ListNode node=listOfHeads.get(i);
            if(node!=null){
                if(minNode==null || node.val<minNode.val){
                    minNode=node;
                    minIndex=i;
                }
            }
        }
        if(minIndex>=0){
            listOfHeads.set(minIndex,minNode.next);
        }
        return minNode;
    }
    
    public ListNode mergeKLists(ArrayList<ListNode> lists) {
        List<ListNode> listOfHeads=new ArrayList<>(lists);
        ListNode vHead=new ListNode(0);
        ListNode cur=vHead,next=null;
        while((next=findMin(listOfHeads))!=null){
            cur.next=next;
            cur=next;
        }
        return vHead.next;
    }
}

File: WeeklyCode/week6/Merge k Sorted Lists/chenhao.py
class Solution:
    # @param a list of ListNode
    # @return a ListNode
    def mergeKLists(self, lists):
        vHead=ListNode(0)
        cur=vHead
        
        heap=[ (listHead.val,listHead) for listHead in lists if listHead!=None ]  
        heapq.heapify(heap) 
        
        while len(heap)!=0:
            val,node=heapq.heappop(heap)
            cur.next=node
            cur=node
            if node.next!=None:
                heapq.heappush(heap,(node.next.val,node.next))
        cur.next=None
        
        return vHead.next